pytorch_lightning/trainer/data_loading.py

```diff
@@ -109,14 +109,49 @@ def auto_add_sampler(self, dataloader: DataLoader, shuffle: bool) -> DataLoader:
 
     def replace_sampler(self, dataloader, sampler):
         skip_keys = ['sampler', 'batch_sampler', 'dataset_kind']
+        skip_valid_keys = ['args', 'kwargs', 'self']
+
+        params = {k:v for k, v in vars(dataloader).items() if not k.startswith("_")}
+
+        valid_kwargs = set(inspect.signature(dataloader.__init__).parameters)
+        contains_dataset = True
+
+        if type(dataloader) is not DataLoader:
+            contains_dataset = "dataset" in valid_kwargs
+            valid_kwargs.update(inspect.signature(DataLoader.__init__).parameters)
 
         dl_args = {
-            k: v for k, v in dataloader.__dict__.items() if not k.startswith('_') and k not in skip_keys
+            name: params[name] for name in valid_kwargs
+            if name in params and name not in skip_keys
         }
-
         dl_args['sampler'] = sampler
         dl_args['shuffle'] = False
+        dl_args['batch_sampler'] = None
```

Comment:
Does this mean that any batch_samplers get replaced? why?

Reply:
great catch ! I added support for it, assuming it follows the same API